### PR TITLE
perf(forms): use `ngDevMode` to tree-shake error messages

### PR DESCRIPTION
**Pull request by sonukapoor**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [X] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #37697

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: packages/forms/src/directives/ng_control.ts [hunk 1]
```diff
@@ -13,7 +13,9 @@ import {ControlValueAccessor} from './control_value_accessor';
 import {AsyncValidator, AsyncValidatorFn, Validator, ValidatorFn} from './validators';
 
 function unimplemented(): any {
-  throw new Error('unimplemented');
+  if (ngDevMode) {
+    throw new Error('unimplemented');
```
Comment (alfaproject) [inline packages/forms/src/directives/ng_control.ts:17]: Might as well not even call this method so that we don't bundle a NOOP method

**Comment by AndrewKushnir** [comment]
FYI, I just did a quick test with a demo app and ReactiveForms module. The `@angular/forms` package got reduced from 41kb to 34kb (which is around 20% decrease) 🎉

We should consider creating a test app with all Forms features (Reactive, Template-driven, validators, etc) in the `integration` folder and have a size check setup there as well, so that we can monitor the effect of the changes performed in different PRs in the `@angular/forms` package size.

File: packages/forms/src/directives/shared.ts [hunk 2]
```diff
@@ -120,13 +120,16 @@ function setUpModelChangePipeline(control: FormControl, dir: NgControl): void {
...
     control: FormGroup|FormArray, dir: AbstractFormGroupDirective|FormArrayName) {
-  if (control == null) _throwError(dir, 'Cannot find control with');
+  if (control == null && ngDevMode) _throwError(dir, 'Cannot find control with');
   control.validator = Validators.compose([control.validator, dir.validator]);
   control.asyncValidator = Validators.composeAsync([control.asyncValidator, dir.asyncValidator]);
 }
 
 function _noControlError(dir: NgControl) {
-  return _throwError(dir, 'There is no FormControl instance attached to form control element with');
+  if (ngDevMode) {
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/shared.ts:129]: I think in this case we can keep the `_noControlError` as is and add the `ngDevMode` check before calling `_noControlError`, so we don't include a noop function.

File: packages/forms/src/directives/radio_control_value_accessor.ts [hunk 3]
```diff
@@ -10,6 +10,7 @@ import {Directive, ElementRef, forwardRef, Injectable, Injector, Input, OnDestro
 
 import {ControlValueAccessor, NG_VALUE_ACCESSOR} from './control_value_accessor';
 import {NgControl} from './ng_control';
+import {RadioControlValueAccessorErrors} from './radio_control_value_accessor_errors';
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/radio_control_value_accessor.ts:13]: We should probably just keep the `throwNameError` function in this file and make it a top level one (i.e. extract from the class).

File: packages/forms/src/directives/reactive_directives/form_control_directive.ts [hunk 4]
```diff
@@ -68,11 +68,13 @@ export class FormControlDirective extends NgControl implements OnChanges {
 
   /**
    * @description
-   * Triggers a warning that this input should not be used with reactive forms.
+   * Triggers a warning in dev mode that this input should not be used with reactive forms.
    */
   @Input('disabled')
   set isDisabled(isDisabled: boolean) {
-    ReactiveErrors.disabledAttrWarning();
+    if (ngDevMode) {
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/reactive_directives/form_control_directive.ts:75]: This is an interesting case: the input is used for dev-mode-only checks, so we should check if we can also tree-shake it away with `ngDevMode` (otherwise we would include a noop function).

File: packages/forms/src/directives/ng_model.ts [hunk 5]
```diff
@@ -317,18 +317,20 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
...
+      if (!(this._parent instanceof NgModelGroup) &&
+          this._parent instanceof AbstractFormGroupDirective) {
+        TemplateDrivenErrors.formGroupNameException();
+      } else if (!(this._parent instanceof NgModelGroup) && !(this._parent instanceof NgForm)) {
+        TemplateDrivenErrors.modelParentException();
+      }
     }
   }
 
   private _checkName(): void {
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/ng_model.ts:330]: It looks like we can make additional refactoring here: this function can be extracted from the class and we can put the `ngDevMode` check before calling it. The following line:
```
  if (this.options && this.options.name) this.name = this.options.name;
```
Can be moved into the main code (before calling the `_checkName` function).

**Review by AndrewKushnir — COMMENTED** [review]
Thanks for these changes @sonukapoor, as I mentioned in my previous comment in the thread, it helps save around 7kb (or ~20%).

I've left a few comments, mainly around additional refactoring that can be made to avoid having noop functions (when the whole body of the function is guarded with `ngDevMode` check).

Another thing that we should consider is to make sure this works correctly with View Engine (since `ngDevMode` was mostly used in Ivy runtime code).

Thank you.

File: packages/forms/src/directives/shared.ts [hunk 6]
```diff
@@ -33,8 +33,10 @@ export function controlPath(name: string|null, parent: ControlContainer): string
 }
 
 export function setUpControl(control: FormControl, dir: NgControl): void {
-  if (!control) _throwError(dir, 'Cannot find control with');
-  if (!dir.valueAccessor) _throwError(dir, 'No value accessor for form control with');
+  if (!control && (typeof ngDevMode === 'undefined' || ngDevMode))
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/shared.ts:36]: For readability, I think we can just wrap 2 `if` statements like this:
```
  if (typeof ngDevMode === 'undefined' || ngDevMode) {
    if (!control) _throwError(dir, 'Cannot find control with');
    if (!dir.valueAccessor) _throwError(dir, 'No value accessor for form control with');
  }
```

File: packages/forms/src/directives/shared.ts [hunk 7]
```diff
@@ -64,8 +66,10 @@ export function setUpControl(control: FormControl, dir: NgControl): void {
 }
 
 export function cleanUpControl(control: FormControl, dir: NgControl) {
-  dir.valueAccessor!.registerOnChange(() => _noControlError(dir));
-  dir.valueAccessor!.registerOnTouched(() => _noControlError(dir));
+  if (typeof ngDevMode === 'undefined' || ngDevMode) {
+    dir.valueAccessor!.registerOnChange(() => _noControlError(dir));
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/shared.ts:70]: In this particular case we can not wrap the logic into the `ngDevMode` check - the `dir.valueAccessor!.registerOnChange` and `dir.valueAccessor!.registerOnTouched` calls should still happen (in both dev and prod modes), since it resets the subscriptions. We can do something like this:
```
  const noop = () => {
    if (typeof ngDevMode === 'undefined' || ngDevMode) {
      _noControlError(dir);
    }
  };
  dir.valueAccessor!.registerOnChange(noop);
  dir.valueAccessor!.registerOnTouched(noop);
```

File: packages/forms/src/directives/ng_model.ts (hunk 5 above)
Comment (AndrewKushnir) [reply packages/forms/src/directives/ng_model.ts:330]: I looked at this code again and I think we can make this refactor later and have only `ngDevMode`-only changes in this PR.

File: packages/forms/src/directives/reactive_directives/form_control_directive.ts [hunk 8]
```diff
@@ -68,11 +68,14 @@ export class FormControlDirective extends NgControl implements OnChanges {
 
   /**
    * @description
-   * Triggers a warning that this input should not be used with reactive forms.
+   * Triggers a warning in dev mode that this input should not be used with reactive forms.
    */
+
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/reactive_directives/form_control_directive.ts:73]: Let's remove this extra line.

File: packages/forms/src/directives/ng_control.ts [hunk 9]
```diff
@@ -67,7 +67,11 @@ export abstract class NgControl extends AbstractControlDirective {
    * @throws An exception that this method is not implemented
    */
   get validator(): ValidatorFn|null {
-    return <ValidatorFn>unimplemented();
+    if (typeof ngDevMode === 'undefined' || ngDevMode) {
+      return <ValidatorFn>unimplemented();
+    } else {
+      return null;
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/ng_control.ts:73]: I think your initial change was actually good (i.e. guard the `throw` with `ngDevMode`). I didn't realize that we actually return from this function and having a return value dependent on `ngDevMode` doesn't look good correct. Could you please revert this change (i.e. move `ngDevMode` back to `unimplemented` function)?

**Comment by AndrewKushnir** [comment]
Hi @sonukapoor, thanks for performing further updates! That's really great to see the dev-mode-only symbols go from the golden file 👍 I just left a quick comment on `unimplemented` function and also [started tests in Google's codebase](http://test/OCL:323096665:BASE:323096698:1595634088430:80a3e8e0). Will keep you updated. Thank you.

File: packages/forms/src/directives/ng_control.ts (hunk 9 above)
Comment (sonukapoor) [reply packages/forms/src/directives/ng_control.ts]: Done.

**Comment by sonukapoor** [comment]
> Hi @sonukapoor, thanks for performing further updates! That's really great to see the dev-mode-only symbols go from the golden file 👍 I just left a quick comment on `unimplemented` function and also [started tests in Google's codebase](http://test/OCL:323096665:BASE:323096698:1595634088430:80a3e8e0). Will keep you updated. Thank you.

No worries. I have added the changes to the same commit.

**Comment by muhammedMoussa** [comment]
> FYI, I just did a quick test with a demo app and ReactiveForms module. The `@angular/forms` package got reduced from 41kb to 34kb (which is around 20% decrease) 🎉
> 
> We should consider creating a test app with all Forms features (Reactive, Template-driven, validators, etc) in the `integration` folder and have a size check setup there as well, so that we can monitor the effect of the changes performed in different PRs in the `@angular/forms` package size.


Is `ngDevMode` mechanism used at other packages likes was in forms, maybe need to be updated as @sonukapoor suggestions, to reduce packages too

**Comment by sonukapoor** [comment]
> Is `ngDevMode` mechanism used at other packages likes was in forms, maybe need to be updated as @sonukapoor suggestions, to reduce packages too

We can definitely look into that as well. @AndrewKushnir might already know more about this.

**Comment by AndrewKushnir** [comment]
Hi @muhammedMoussa, the `ngDevMode` flag is actively used in Ivy runtime code to tree-shake away all dev-mode-only code in production builds. In this PR we are trying out this approach with Forms package. If everything goes well and there are no major blockers, we would consider doing the same in other packages too. Thank you.

File: packages/forms/src/directives/ng_control.ts [hunk 10]
```diff
@@ -13,7 +13,9 @@ import {ControlValueAccessor} from './control_value_accessor';
 import {AsyncValidator, AsyncValidatorFn, Validator, ValidatorFn} from './validators';
 
 function unimplemented(): any {
-  throw new Error('unimplemented');
+  if (typeof ngDevMode === 'undefined' || ngDevMode) {
```
Comment (mhevery) [inline packages/forms/src/directives/ng_control.ts:16]: I don't believe that calling `typeof ngDevMode === 'undefined'` is necessary. As long as angular code runs first it should be properly set up. The only place this would be an issue is if you used `ngDevMode` before call to angular (ie from the top level file context) So. I don't think it is necessary here.

**Review by mhevery — APPROVED** [review]
LGTM, 

but see my comment about `typeof ngDevMode === 'undefined'` (I don't think it is needed)

File: packages/forms/src/directives/ng_control.ts (hunk 10 above)
Comment (AndrewKushnir) [reply packages/forms/src/directives/ng_control.ts:16]: Thanks for the review Misko! 👍 

This code needs to run with both Ivy and View Engine and if I understand correctly, the `ngDevMode` would not be initialized in VE (thus there is a risk of having `ngDevMode` undeclared here)?

**Comment by mhevery** [comment]
I see, yes I think you are right there.

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Review by clydin — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]

**Comment by clydin** [comment]
By using the example forms code found [here](https://github.com/angular/angular/blob/763023472b15f82a7551c9c4f8f43cda1d259dd5/packages/core/test/bundling/forms/index.ts) in combination with a newly generated CLI application, a savings of ~7kb was observed for both View Engine and Ivy renderers.

One area that might be beneficial to investigate is to potentially change the View Engine startup process to guarantee that `ngDevMode` is defined.  That should allow the removal of the `undefined` checks throughout this PR.

**Comment by AndrewKushnir** [comment]
[Presubmit #2](http://test/OCL:326349299:BASE:326349329:1597278786661:ab80e26a).

**Comment by AndrewKushnir** [comment]
FYI, tests in Google's codebase (global presubmit) went well.

Also adding a link to similar changes in Components repo for visibility: https://github.com/angular/components/pull/20146.
